File: dsp18.pas
(********************  PROGRAM NO.18  **************************)
(*****  NONRECURSIVE FILTER DESIGN WITH THE KAISER WINDOW  *****)
program DSP18;
uses graph3;
var SEE: array[0..200] of real;
var H: array[0..200] of real;
var W: array[0..320] of real;
var P,Q,R,S,T,A,I,B,C,D: real;
var ALPHA,BETA,A0,H0,W0,W1,FREQ,MAX,DB,MREAL: real;
var X,N,K,M,Y: integer;
label 1,2;
begin
  1:writeln('ENTER CENTER-FREQUENCY (IN DEGREES)'); readln(P);
    W0:=P*3.141593/180;
    writeln('ENTER FILTER BANDWIDTH (IN DEGREES)'); readln(Q);
    W1:=Q*0.5*3.141593/180;
    writeln('ENTER KAISER DESIGN PARAMETERS:');
    writeln('RIPPLE (AS A FRACTION):'); readln(R);
    writeln('TRANSITION WIDTH (DEGREES):'); readln(S); T:=S/360;
(***************  ESTIMATE KAISER PARAMETERS  ******************)
 A:=-20*ln(R)*0.4343;MREAL:=(A-7.95)/(28.72*T);M:=trunc(MREAL+1);
  writeln('NO. OF IMPULSE RESPONSE TERMS=',2*M+1);
  writeln('CHANGE  PARAMETERS, OR CONTINUE (0/1)?');
  readln(X); if X=0 then GOTO 1;
    if A>49 then begin ALPHA:=0.1102*(A-8.7);goto 2 end;
    if A>21 then begin
     ALPHA:=(0.5842*exp(0.4*ln(A-21))+(0.07886*(A-21)));goto 2
    end;ALPHA:=0;
  2:writeln('FOR THIS WINDOW, ALPHA=',ALPHA);
    writeln('ENTER ANY INTEGER TO START PLOT'); readln(X);
(*******************  COMPUTE WINDOW VALUES  *******************)
  for N:=0 to M do begin
    BETA:=ALPHA*sqrt(1-(N/M)*(N/M));I:=1;B:=1;
      for K:=1 to 20 do begin
        C:=B*BETA*BETA/4;D:=C/(K*K);I:=I+D;B:=D;end;SEE[N]:=I;
  end;A0:=SEE[0]; for N:=0 to M do begin
                    SEE[N]:=SEE[N]/A0;end;
(**************  COMPUTE IMPULSE RESPONSE VALUES  **************)
  H0:=W1/3.141593;
    for N:=1 to M do begin
      H[N]:=(1/(N*3.141593))*sin(N*W1)*cos(N*W0)*SEE[N];end;
(**************  DRAW RECTANGULAR GRID FOR PLOT  ***************)
  graphmode;
    for K:=0 to 5 do begin draw(0,30+20*K,320,30+20*K,1);
             draw(64*K,30,64*K,130,1);end;draw(319,30,319,130,1);
(***************  COMPUTE FREQUENCY RESPONSE  ******************)
  for N:=1 to 320 do begin
    FREQ:=3.141593*(N-1)/320;W[N]:=H0;
      for K:=1 to M do begin
        W[N]:=W[N]+2*H[K]*cos(K*FREQ);end;end;
(*****  NORMALISE TO UNITY, CONVERT TO DECIBELS, AND PLOT  *****)
  MAX:=0;
  for N:=1 to 320 do begin
    if abs(W[N])>MAX then MAX:=abs(W[N]);end;
      for N:=1 to 320 do begin
        DB:=20*ln(abs(W[N])/MAX)*0.4343;if DB<-50 then DB:=-50;
        draw(N,130,N,round(30-2*DB),1);end;
    writeln('ENTER ANY INTEGER TO PRINT h[n] VALUES');readln(Y);
  clearscreen;
    writeln('VALUES h[0] TO h[M]:');
    writeln('(CORRECTED FOR UNITY MAXIMUM GAIN)');
    writeln(H0/MAX);
  for N:=1 to M do begin writeln(H[N]/MAX);end;
end.
(***************************************************************)

File: dsp18.c
#include <graph.h>
#include <conio.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define SEEDIM    200
#define HDIM      200
#define WDIM      320

float fSEE[SEEDIM + 1], fH[HDIM + 1], fW[WDIM + 1];

void main (void)
    {
    /*
     ******************* PROGRAM NO.18 *************************
     **** NONRECURSIVE FILTER DESIGN WITH THE KAISER WINDOW ****
     */
    int    iCount, iCount2, iX, iM;
    float  fM, fP, fQ, fR, fS, fT, fA, fA0, fB, fC, fD,
           fALPHA, fBETA, fTEMP, fI,
           fH0, fW0, fW1, fFREQ, fMax, fDB;

    _clearscreen  (_GCLEARSCREEN);
    printf ("PROGRAM 18: Nonrecursive filter design with\n");
    printf ("the Kaiser window\n\n");

    while (1)
        {
        printf ("Enter center-frequency (in degrees) :");
        scanf ("%f", &fP);

        printf ("Enter filter bandwidth (in degrees) :");
        scanf ("%f", &fQ);

        printf ("Enter Kaiser design parameters\n");
        printf ("  Ripple (as a fraction) :");
        scanf ("%f", &fR);

        printf ("\nTransition width (in degrees) :");
        scanf ("%f", &fS);

        fW0 = fP * 3.141593 / 180.0;
        fW1 = fQ * 0.5 * 3.141593 / 180.0;
        fT = fS / 360.0;

        /*
         *************** ESTIMATE KAISER PARAMETERS ************
         */
        fA = -20.0 * log (fR) * 0.4343;
        fM = (fA - 7.95) / (28.72 * fT);
        iM = ((int)(fM+1));

        printf ("\n  No. of impulse response terms = %d\n",
                              2 * iM + 1);
        printf ("\n  Change Parameters, or continue? (0/1) :");
        scanf ("%d", &iX);

        if (iX)
            break;
        }

    if (fA > 49.0)
        fALPHA = 0.1102 * (fA - 8.7);
    else if (fA > 21)
        fALPHA = 0.5842 * pow((fA - 21.0), 0.4) +
                 (0.07886 * (fA - 21.0));
    else
        fALPHA = 0;

    printf ("\nFor this window, ALPHA = %f\n", fALPHA);
    printf ("Press any key to start...");
    getch();

    /*
     ******************* COMPUTE WINDOW VALUES *****************
     */
    for (iCount = 0; iCount <= iM; iCount++)
        {
        fTEMP = 1.0 - pow (((float) iCount / (float) iM), 2.0);
        fBETA = fALPHA * pow (fTEMP, 0.5);
        fI = fB = 1.0;

        for (iCount2 = 1; iCount2 <= 20; iCount2++)
            {
            fC = fB * fBETA * fBETA / 4.0;
            fD = fC / (float)(iCount2 * iCount2);
            fI += fD;
            fB = fD;
            }

        fSEE[iCount] = fI;
        }

    fA0 = fSEE[0];
    for (iCount = 0; iCount <= iM; iCount++)
        fSEE[iCount] /= fA0;

    /*
     *************** COMPUTE IMPULSE RESPONSE VALUES ***********
     */
    fH0 = fW1 / 3.141593;

    for (iCount = 1; iCount <= iM; iCount++)
        fH[iCount] = (1.0 / ((float)iCount * 3.141593)) *
                      sin ((float)iCount * fW1) *
                      cos ((float)iCount * fW0) *
                      fSEE[iCount];

    _setvideomode( _MRES16COLOR );

    /*
     ************** DRAW RECTANGULAR GRID FOR PLOT *************
     */
    for (iCount = 0; iCount <= 5; iCount++)
        {
        _moveto (0, 30 + 20 * iCount);
        _lineto (320, 30 + 20 * iCount);
        _moveto (64 * iCount, 30);
        _lineto (64 * iCount, 130);
        }

    _moveto (319, 30);
    _lineto (319, 130);

    /*
     **************** COMPUTE FREQUENCY RESPONSE ***************
     */
    for (iCount = 1; iCount <= WDIM; iCount++)
        {
        fFREQ = 3.141593 * (float)(iCount - 1) / 320.0;
        fW[iCount] = fH0;
        for (iCount2 = 1; iCount2 <= iM; iCount2++)
            fW[iCount] += 2.0 * fH[iCount2] *
                         cos ((float)iCount2 * fFREQ);
        }

    /*
     ****** NORMALISE TO UNITY, CONVERT TO DECIBELS & PLOT *****
     */
    fMax = 0;

    for (iCount = 1; iCount <= WDIM; iCount++)
        if (fabs (fW[iCount]) > fMax)
            fMax = fabs (fW[iCount]);

    for (iCount = 1; iCount <= WDIM; iCount++)
        {
        fDB = 20.0 * log (fabs (fW[iCount]) / fMax) * 0.4343;

        if (fDB < -50.0)
            fDB = -50.0;

        _moveto (iCount, 130);
        _lineto (iCount, (int)(30 - 2.0 * fDB + 0.5));
        }

    printf ("Press any key to continue...");
    getch();

    _setvideomode( _DEFAULTMODE );

    printf ("Impulse response values h(0) to h(M)\n");
    printf ("(corrected for unity maximum gain) :\n\n");

    printf ("%18.5f", (fH0 / fMax));

    for (iCount = 0; iCount < iM; iCount++)
        printf (", %18.5f", (fH[iCount] / fMax));

    /*
     ***********************************************************
     */
    }
